added under command Issue #2331 (#2354)
python-discord/bot: bot/resources/tags/under.md

**Use cases of Underscore in Python**

- **Dunders**: Double Underscore (Dunder) methods are special methods defined in a class that are invoked implicitly. For example, `__name__`,`__init__`,`__repr__`, and `__str__`.
Use the `!dunder-methods` tag to know more about these methods
- **Single leading Underscore**:  In front of a variable, a function, or a method name means that these objects are used internally, such as `_name`. Remember that these objects can be accessed outside or other script.

> - **Single leading Underscore**:  In front of a variable, a function, or a method name means that these objects are used internally, such as `_name`. However, remember that these objects **can** be accessed outside.

Reviewer: Seems you have the "Single leading underscore" duplicated.
I think you were referring to "Single trailing underscore", right ? Description needs fixing as well

> - **Single leading Underscore**:  In front of a variable, a function, or a method name means that these objects are used internally, such as `_name`. However, remember that these objects **can** be accessed outside.

Reviewer: I think it would be better if we phrase it like this:
These mean that they're used in front a variable, a function or a method name, which indicates that these objects are meant to be used internally, such as `_name`. However, these objects **can still** be accessed from outside

We could also benefit from saying that the usage of these conventions is to remediate to the lack of "access modifiers" in Python, we'll let others give their opinion on this as well.

-  **Double leading underscores**: Double leading underscores are typically used for name mangling.

> Name mangling is a process by which the interpreter changes the attribute name to avoid naming collisions in subclasses.[GeeksforGeeks](https://www.geeksforgeeks.org/name-mangling-in-python/)

Reviewer: I think  you should say something like: More details & examples about name mangling can be found [here](https://www.geeksforgeeks.org/name-mangling-in-python/).

> - **Store Expression Value** Underscore(**_**) is used to store the value of last expression in an interpreter and can be used as a variable. Example:

Reviewer: You still didn't explain that if this variable has been assigned once, it won't allow the latest evaluated value from the REPL anymore.
Example
```python
>>> 1 + 1
2
>>> 2 + _
4
>>> _
4
>>> _ = 12
>>> 1 + 1
2
>>> 2 + _
14
>>>
```

```python
>>> 1 + 1
    2
>>> _
    2
>>> _ + 2
    4
```

> -  **Ignore Values**: Underscores can also be used for variable assignments, but they are generally used to store values that will not be needed.It is known as `throwaway variable`

Reviewer: ```suggestion
-  **Ignore Values**: Underscores can also be used for variable assignments, but they are generally used that way to store values that will not be needed. It is known as `throwaway variable`
```

> ## UseCases of Underscore in Python

Reviewer 2: This doesn't work in Discord embeds. It'll show up as the following:
<img width="251" alt="image" src="https://user-images.githubusercontent.com/15021300/206897545-1beb67e8-aa9c-48a3-9998-0c504350bfa3.png">

```suggestion
**UseCases of Underscore in Python**

```

> - **Dunders** Double Underscore(Dunder) methods are special methods defined in a class that are invoked implicitly. example- `__name__`,`__init__`,`__repr__`,`__str__`

Reviewer 2: ```suggestion
- **Dunders**: Double Underscore (Dunder) methods are special methods defined in a class that are invoked implicitly. For example, `__name__`,`__init__`,`__repr__`, and `__str__`.
```

Use `!dunder-methods` command to know more about these methods

> - **Single leading Underscore**  in front of a variable, a function, or a method name means that these objects are used internally.Remember that these objects can be accessed outside or other script. example-`_name`

Reviewer 2: ```suggestion
- **Single leading Underscore**:  In front of a variable, a function, or a method name means that these objects are used internally, such as `_name`. Remember that these objects can be accessed outside or other script. 
```

> - **Single Trailing Underscores** are used to avoid naming conflict while using a variable which is a reserved keyword.
> example- using a variable name as `class` will produce an error,to avoid this conflict, you can add a trailing underscore as a naming convention i.e. `class_`.

Reviewer 2: I think this should be on a single line.

```suggestion
- **Single Trailing Underscores**: Avoid naming conflict while using a variable which is a reserved keyword. Using a variable name as `class` will produce an error. To avoid this conflict, you can add a trailing underscore as a naming convention i.e. `class_`.
```

> -  **Double leading underscores** Double leading underscores are typically used for name mangling.

Reviewer 2: ```suggestion
-  **Double leading underscores**: Double leading underscores are typically used for name mangling.
```

Name mangling is a process by which the interpreter changes the attribute name to avoid naming collisions in subclasses.

> - **Store Expression Value** Underscore(**_**) is used to store the value of last expression in an interpreter and can be used as a variable.example-

Reviewer 2: I don't think it's necessary to say that we're using an underscore, since that is what this tag is about.

```suggestion
- **Store Expression Value**: Store the value of last expression in an interpreter and can be used as a variable.
```

```python

> >>>1+1

Reviewer 2: Keep the spacing consistent.

```suggestion
>>> 1 + 1
```

    2
>>> _
    2
>>> _ + 2
    4
```

> -  **Ignore Values** Underscore can be used to ignore values and as a variable in looping.example-

Reviewer 2: ```suggestion
-  **Ignore Values**: An underscore can be used to ignore values and as a variable in looping.
```

 ```python

> >>> a, *_, b = (3,5,2,6,2)
> >>> print(a,b,_)

Reviewer 2: Let's keep spaces between commas consistent.

```suggestion
>>> a, *_, b = (3, 5, 2, 6, 2)
>>> print(a, b, _)
```

     3 2 [5, 2, 6]

> >>> for _ in range(2):  #variable in looping

Reviewer 2: There should be a space after the `#`.

```suggestion
>>> for _ in range(2):  # variable in looping
```

          print(_)
       0
       1
```

> -  **Improve Readability** - can be used with numeric literals to improve readability of long numbers.example-

Reviewer 2: ```suggestion
-  **Improve Readability**: they can also be used with numeric literals to improve readability of long numbers.
```

 ```python

> >>>1_00_000

Reviewer 2: ```suggestion
>>> 1_00_000
```

    100000
```

> References-

Reviewer 2: ```suggestion

**References**:
```

> **UseCases of Underscore in Python**

Reviewer 3: ```suggestion
**Use cases of Underscore in Python**
```

- **Dunders**: Double Underscore (Dunder) methods are special methods defined in a class that are invoked implicitly. For example, `__name__`,`__init__`,`__repr__`, and `__str__`.

> Use `!dunder-methods` command to know more about these methods

Reviewer 4: ```suggestion
Use the `!dunder-methods` tag to know more about these methods
```

> - **Single leading Underscore**:  In front of a variable, a function, or a method name means that these objects are used internally, such as `_name`. Remember that these objects can be accessed outside or other script.

Reviewer 4: ```suggestion
- **Single leading Underscore**:  In front of a variable, a function, or a method name means that these objects are used internally, such as `_name`. However, remember that these objects **can** be accessed outside.
```

- **Single Trailing Underscores**: Avoid naming conflict while using a variable which is a reserved keyword. Using a variable name as `class` will produce an error. To avoid this conflict, you can add a trailing underscore as a naming convention i.e. `class_`.
-  **Double leading underscores**: Double leading underscores are typically used for name mangling.

> Name mangling is a process by which the interpreter changes the attribute name to avoid naming collisions in subclasses.

Reviewer: I think it would be helpful to link some article that explains it better here

> - **Store Expression Value** Underscore(**_**) is used to store the value of last expression in an interpreter and can be used as a variable.example-

Reviewer 4: ```suggestion
- **Store Expression Value** Underscore(**_**) is used to store the value of last expression in an interpreter and can be used as a variable. Example:
```

> - **Store Expression Value** Underscore(**_**) is used to store the value of last expression in an interpreter and can be used as a variable.example-

Reviewer: It is indeed correct that it can be used as a variable, but I think we should state that we usually only use it for variables that are not needed from return statements, iterations, etc.
I know you've mentioned later on, but here's how I think it should be
"Underscores can also be used for variable assignments, but they are generally used to store values that will not be needed."
Then you can give your detailed examples (i see you do that later down the tag) where you say how they're using in return statements, and maybe in a loop, but stating the **context** for each example.
We should always keep in mind that these examples need to be as understandable as they can be, and we especially want to reduce frustrations for total beginners.
Let me know if it's not clear enough and I'll gladly assist you

> - **Store Expression Value** Underscore(**_**) is used to store the value of last expression in an interpreter and can be used as a variable.example-

Reviewer: Maybe "python shell" is better than interpreter here ? We can wait for other people's opinions here as well

```python
>>> 1 + 1
    2
>>> _
    2
>>> _ + 2
    4
```
-  **Ignore Values**: An underscore can be used to ignore values and as a variable in looping.
 ```python
>>> a, *_, b = (3, 5, 2, 6, 2)
>>> print(a, b, _)
     3 2 [5, 2, 6]
>>> for _ in range(2):  # variable in looping

>           print(_)

Reviewer: Looks like you have one extra indentation level here, can you fix that please ?

**Use cases of Underscore in Python**

- **Dunders**: Double Underscore (Dunder) methods are special methods defined in a class that are invoked implicitly. For example, `__name__`,`__init__`,`__repr__`, and `__str__`.

> Use the `!dunder-methods` tag to know more about these methods.

Reviewer: ```suggestion
Use the `!dunder-methods` tag to know more about them.
```

> - **Single leading Underscore**:  They're used in front a variable, a function or a method name, which indicates that these objects are meant to be used internally, such as `_name`. However, these objects **can** still be accessed from outside.

Reviewer: Nit: Single Leading Underscores

> - **Single Trailing Underscores**: They are used to avoid naming conflict while using a variable, which is a reserved keyword. Using a variable name as `class` will produce an error. To avoid this conflict, you can add a trailing underscore as a naming convention i.e. `class_`.

Reviewer: ```suggestion
- **Single Trailing Underscores**: They are used to avoid naming conflict while making a new variable whose name might conflict with a reserved keyword or package name, etc. For example, using `class` as a variable name will produce an error; and to avoid this conflict, you can add a trailing underscore to it, i.e. `class_`.
```

> -  **Double leading underscores**: Double leading underscores are typically used for Name Mangling. Name mangling is a process by which the interpreter changes the attribute name to avoid naming collisions in subclasses. More details regarding Name Mangling can be found [here](https://www.geeksforgeeks.org/name-mangling-in-python/).

Reviewer: ```suggestion
-  **Double Leading Underscores**: Double leading underscores are typically used for "Name Mangling". Name mangling is a process by which the interpreter changes the attribute name to avoid naming collisions in subclasses. More details regarding this topic can be found [here](https://www.geeksforgeeks.org/name-mangling-in-python/).
```

- **Store Expression Value**: Underscore(**_**) is used to store the value of last expression in an interpreter and can be used as a variable. Example:
```python
>>> _ # Using underscore without any previously executed expression or value assigned to underscore
    NameError: name '_' is not defined
>>> 1 + 1  # evaluating expression
    2

> >>> _        # prints value of last executed expression

Reviewer: ```suggestion
>>> _   # Prints the value of the last executed expression
```

    2
>>> _ + 2
    4

> >>> _ = 3  # Once assigned any value,Underscore doesn't store the value of last executed expression anymore

Reviewer: ```suggestion
>>> _ = 3  # Once a value has explicitly been assigned to it, underscore won't store the value of last executed expression anymore.
```

>>> 5 + 9
    14
>>> _
    3
```

> -  **Ignore Values**: Underscores can also be used for variable assignments, but they are generally used that way to store values that will not be needed. It is known as `throwaway variable`.

Reviewer: ```suggestion
-  **Ignore Values**: We previously stated that underscores can also be used for variable assignments, but they are most commonly used that way to store values that will not be needed. It is known as `throwaway variable`.
```

 ```python
>>> a, *_, b = (3, 5, 2, 6, 2) # Ignore values when unpacking
>>> print(a, b, _)
    3 2 [5, 2, 6]

> >>> for _ in range(10):  # Ignoring the Index

Reviewer: ```suggestion
>>> for _ in range(10):  # Here the index is ignored since we only want to execute the same function multiple times
```
